packages/async-flow/src/log-store.js

```diff
@@ -113,19 +149,48 @@ export const prepareLogStore = zone => {
         const result = mapStore.get(eph.index);
         return result;
       },
-      nextEntry() {
+      /**
+       * @returns {LogEntry}
+       */
+      nextUnfilteredEntry() {
         const { self } = this;
         const eph = tmp.for(self);
 
         const result = self.peekEntry();
         eph.index += 1;
+        if (entryIsVisible(result)) {
+          eph.filteredIndex += 1;
+        }
         if (!self.isReplaying()) {
           eph.replayDoneKit.resolve(undefined);
         }
         return result;
       },
+      /**
+       * @returns {LogEntry}
+       */
+      nextEntry() {
+        const { self } = this;
+        let result = self.nextUnfilteredEntry();
+        while (self.isReplaying() && !entryIsVisible(result)) {
+          result = self.nextUnfilteredEntry();
+        }
```

Comment:
Technically the `self.isReplaying()` is redundant since a `startGeneration` entry would always appear before another entry in the log.

I suppose we may have other invisible entries in the future that are at the tail of the log, but I think that would mess other cases of "is replaying check", so that may be something we likely want to prevent anyways.

```suggestion
        while (!entryIsVisible(result)) {
          self.isReplaying() || Fail`Unexpected entry at log tail: ${result}`;
          result = self.nextUnfilteredEntry();
        }
```

Reply:
Committed your suggestion.